M1 Studio qual: Python backend health check fails during dev-up (Connection refused on port 10464)

## Summary

macOS beta qualification consistently fails at `desktop-preparation` phase because the Python backend starts (process is alive) but never passes its HTTP health check within the dev-up window. All other services (Firebase, Redis, Typesense, desktop-backend) pass fine.

## Evidence

**Affected runs (all same failure class):**
- [v0.12.144+12144-macos](https://github.com/BasedHardware/omi/actions/runs/30419409562) — backend pid 10413 started, `Connection refused` on port 10464
- [v0.12.143+12143-macos](https://github.com/BasedHardware/omi/actions/runs/30412843790) — same pattern
- [v0.12.142+12142-macos](https://github.com/BasedHardware/omi/actions/runs/30333312093) — same pattern

**Log signature:**
```
backend: started pid <N>
...
Health checks failed:
  - backend: <urlopen error [Errno 61] Connection refused>
make: *** [dev-up] Error 1
qualification failed: desktop launch process exited during preparation
```

**What works:** Codemagic build fully green, smoke test all 14 checks green, artifact is good. This is purely a runner/harness startup race.

## Likely area

`scripts/dev-harness/dev-up.sh` — the backend health-check polling loop or the backend startup sequence itself. The backend process binds to port 10464 but either:
1. Takes too long to start accepting connections (health check timeout too short)
2. Fails silently during import/init before binding
3. Has a port conflict or bind race with another process

## Acceptance criteria

- [ ] M1 Studio `desktop_qualify_beta.yml` reaches Tier-2/fault suite (past dev-up) consistently
- [ ] Fix includes a regression test or guard so this class of startup race does not silently return

--- a/scripts/dev-harness/dev_harness/cli.py
+++ b/scripts/dev-harness/dev_harness/cli.py
@@ -620,7 +620,15 @@ def _typesense_command(cfg: config.HarnessConfig) -> list[str]:
     ]
 
 
-def _start_services(cfg: config.HarnessConfig) -> None:
+# Infrastructure services (firestore, auth, redis, typesense) start first so the
+# Python backend can connect to them immediately on boot.  The brief settle delay
+# prevents a port-binding race on resource-constrained runners (e.g. M1 Studio
+# under qualification load) where the backend starts before the emulator has
+# finished binding its port.
+_INFRA_SETTLE_DELAY = 2.0
+
+
+def _start_infrastructure(cfg: config.HarnessConfig) -> None:
     cfg.layout.logs_dir.mkdir(parents=True, exist_ok=True)
     _start_process(
         cfg,
@@ -662,6 +670,10 @@ def _start_services(cfg: config.HarnessConfig) -> None:
         log_name="typesense.log",
         port=cfg.typesense_port,
     )
+
+
+def _start_app_services(cfg: config.HarnessConfig) -> None:
+    """Start backend and desktop-backend after infrastructure is settling."""
     _start_process(
         cfg,
         "backend",
@@ -681,7 +693,42 @@ def _start_services(cfg: config.HarnessConfig) -> None:
     )
 
 
-def _wait_health(cfg: config.HarnessConfig, *, timeout: float = 45.0) -> list[str]:
+def _start_services(cfg: config.HarnessConfig) -> None:
+    _start_infrastructure(cfg)
+    # Give infrastructure services a brief head start so the Python backend can
+    # bind connections to Redis, Firestore, and Typesense immediately on boot.
+    # Without this, on a loaded runner the backend may retry connections during
+    # its startup window, extending boot time beyond the health-check deadline.
+    time.sleep(_INFRA_SETTLE_DELAY)
+    _start_app_services(cfg)
+
+
+# Per-service health-check deadlines (seconds).  The Python backend needs the
+# longest window: it imports heavy ML/NLP modules and initialises connections to
+# every infrastructure service.  On an M1 Studio runner under qualification load,
+# 45 s (the old flat deadline shared across *all* services) is not enough.
+_HEALTH_TIMEOUTS: dict[str, float] = {
+    "firestore": 45.0,
+    "auth": 45.0,
+    "typesense": 45.0,
+    "backend": 90.0,
+    "desktop-backend": 60.0,
+    "redis": 30.0,
+}
+
+
+def _wait_health(
+    cfg: config.HarnessConfig,
+    *,
+    timeout: float | None = None,
+) -> list[str]:
+    """Wait for all harness services to become healthy.
+
+    Each service has its own deadline (see ``_HEALTH_TIMEOUTS``) measured from
+    the moment the wait begins.  If a recorded process dies before its deadline,
+    the service is marked unhealthy immediately instead of polling uselessly.
+    Pass ``timeout`` to override the backend deadline for backwards-compat callers.
+    """
     typesense_headers = {"X-TYPESENSE-API-KEY": config.LOCAL_TYPESENSE_API_KEY}
     checks = {
         "firestore": (f"http://{cfg.firestore_host}/", None),
@@ -689,12 +736,44 @@ def _wait_health(cfg: config.HarnessConfig, *, timeout: float = 45.0) -> list[st
         "typesense": (f"http://127.0.0.1:{cfg.typesense_port}/collections", typesense_headers),
         "backend": (f"{cfg.backend_url}/docs", None),
         "desktop-backend": (f"{cfg.desktop_backend_url}/health", None),
+        "redis": (None, None),  # port-based check
     }
     pending = dict(checks)
-    deadline = time.time() + timeout
+    start = time.time()
+    # Per-service deadlines; ``timeout`` overrides the backend deadline only.
+    deadlines: dict[str, float] = {}
+    for service in pending:
+        base = _HEALTH_TIMEOUTS.get(service, 45.0)
+        if timeout is not None and service == "backend":
+            base = timeout
+        deadlines[service] = start + base
     failures: dict[str, str] = {}
-    while pending and time.time() < deadline:
+    process_records = {r["service"]: r for r in _process_records(cfg)}
+    while pending:
+        now = time.time()
+        # Expire services whose per-service deadline has passed.
+        for service in list(pending):
+            if now >= deadlines[service]:
+                url = pending[service][0]
+                failures.setdefault(service, f"not healthy after {deadlines[service] - start:.0f}s at {url}")
+                pending.pop(service)
+        if not pending:
+            break
         for service, (url, headers) in list(pending.items()):
+            # Fail fast if the process died — no point polling a dead service.
+            record = process_records.get(service)
+            if record:
+                pid_val = int(record.get("pid", -1))
+                if not safety.process_exists(pid_val):
+                    failures[service] = f"process exited (pid={pid_val}); check log: {record.get('log', '?')}"
+                    pending.pop(service)
+                    print(f"{service}: {failures[service]}")
+                    continue
+            if service == "redis":
+                if _port_open("127.0.0.1", cfg.redis_port):
+                    print("redis: healthy (port-open)")
+                    pending.pop(service)
+                continue
             ok, detail = _http_ok(url, headers=headers)
             if ok:
                 print(f"{service}: healthy ({detail})")
@@ -705,7 +784,7 @@ def _wait_health(cfg: config.HarnessConfig, *, timeout: float = 45.0) -> list[st
             time.sleep(0.75)
     for service, (url, _) in pending.items():
         failures.setdefault(service, f"not healthy at {url}")
-    return [f"{service}: {failures.get(service, 'unknown failure')}" for service in pending]
+    return [f"{service}: {failures.get(service, 'unknown failure')}" for service in pending] if failures else []
 
 
 def cmd_up(args: argparse.Namespace) -> int:

